src/textual/widgets/_data_table.py

```diff
@@ -26,40 +29,114 @@
 from ..scroll_view import ScrollView
 from ..strip import Strip
 
+CellCacheKey: TypeAlias = "tuple[RowKey, ColumnKey, Style, bool, bool, int]"
+LineCacheKey: TypeAlias = (
+    "tuple[int, int, int, int, Coordinate, Coordinate, Style, CursorType, bool, int]"
+)
+RowCacheKey: TypeAlias = (
+    "tuple[RowKey, int, Style, Coordinate, Coordinate, CursorType, bool, bool, int]"
+)
 CursorType = Literal["cell", "row", "column", "none"]
-CELL: CursorType = "cell"
 CellType = TypeVar("CellType")
 
 
 class CellDoesNotExist(Exception):
-    pass
+    """Raised when the user supplies coordinates or cell keys which
+    do not exist in the DataTable."""
 
 
-def default_cell_formatter(obj: object) -> RenderableType | None:
-    """Format a cell in to a renderable.
+class DuplicateKey(Exception):
+    """Raised when the RowKey or ColumnKey provided already refers to
+    an existing row or column in the DataTable. Keys must be unique."""
+
+
+@functools.total_ordering
+class StringKey:
+    """An object used as a key in a mapping. It can optionally wrap a string,
+    and lookups into a map using the object behave the same as lookups using
+    the string itself."""
+
+    value: str | None
+
+    def __init__(self, value: str | None = None):
+        self.value = value
+
+    def __hash__(self):
+        # If a string is supplied, we use the hash of the string. If no string was
+        # supplied, we use the default hash to ensure uniqueness amongst instances.
+        return hash(self.value) if self.value is not None else id(self)
+
+    def __eq__(self, other: object) -> bool:
+        # Strings will match Keys containing the same string value.
+        # Otherwise, you'll need to supply the exact same key object.
+        if isinstance(other, (str, StringKey)):
+            return hash(self) == hash(other)
+        else:
+            raise NotImplemented
+
+    def __lt__(self, other):
+        if isinstance(other, str):
+            return self.value < other
+        elif isinstance(other, StringKey):
+            return self.value < other.value
+        else:
+            raise NotImplemented
+
+    def __rich_repr__(self):
+        yield "value", self.value
+
+
+class RowKey(StringKey):
+    """Uniquely identifies a row in the DataTable. Even if the visual location
+    of the row changes due to sorting or other modifications, a key will always
+    refer to the same row."""
+
+
+class ColumnKey(StringKey):
+    """Uniquely identifies a column in the DataTable. Even if the visual location
+    of the column changes due to sorting or other modifications, a key will always
+    refer to the same column."""
```

Comment:
May want reformatting to confirm with [our docstring guidelines](https://github.com/Textualize/textual/wiki/Documentation-standards#class).